packages/queued-request-controller/src/QueuedRequestController.test.ts

```diff
@@ -33,208 +32,500 @@ describe('QueuedRequestController', () => {
-        {
-          ...buildRequest(),
-          networkClientId: 'differentNetworkClientId',
-        },
+        buildRequest(),
         () => new Promise((resolve) => setTimeout(resolve, 10)),
       );
 
       expect(mockSetActiveNetwork).toHaveBeenCalledWith(
         'differentNetworkClientId',
       );
+      expect(onNetworkSwitched).toHaveBeenCalledWith(
+        'differentNetworkClientId',
+      );
     });
 
-    it('does not switch networks if a request comes in for the same chain', async () => {
+    it('does not switch networks if a request comes in for the same network client', async () => {
       const mockSetActiveNetwork = jest.fn();
       const { messenger } = buildControllerMessenger({
         networkControllerGetState: jest.fn().mockReturnValue({
           ...cloneDeep(defaultNetworkState),
           selectedNetworkClientId: 'selectedNetworkClientId',
         }),
         networkControllerSetActiveNetwork: mockSetActiveNetwork,
+        selectedNetworkControllerGetNetworkClientIdForDomain: jest
+          .fn()
+          .mockImplementation((_origin) => 'selectedNetworkClientId'),
       });
+      const onNetworkSwitched = jest.fn();
+      messenger.subscribe(
+        'QueuedRequestController:networkSwitched',
+        onNetworkSwitched,
+      );
       const options: QueuedRequestControllerOptions = {
         messenger: buildQueuedRequestControllerMessenger(messenger),
       };
       const controller = new QueuedRequestController(options);
 
       await controller.enqueueRequest(
-        {
-          ...buildRequest(),
-          networkClientId: 'selectedNetworkClientId',
-        },
+        buildRequest(),
         () => new Promise((resolve) => setTimeout(resolve, 10)),
       );
 
       expect(mockSetActiveNetwork).not.toHaveBeenCalled();
+      expect(onNetworkSwitched).not.toHaveBeenCalled();
     });
 
-    it('does not switch networks if the switch chain confirmation is rejected', async () => {
-      const mockSetActiveNetwork = jest.fn();
-      const { messenger } = buildControllerMessenger({
-        approvalControllerAddRequest: jest
-          .fn()
-          .mockRejectedValue(new Error('Rejected')),
-        networkControllerGetState: jest.fn().mockReturnValue({
-          ...cloneDeep(defaultNetworkState),
-          selectedNetworkClientId: 'selectedNetworkClientId',
-        }),
-        networkControllerSetActiveNetwork: mockSetActiveNetwork,
-      });
+    it('queues request if a request from another origin is being processed', async () => {
       const options: QueuedRequestControllerOptions = {
-        messenger: buildQueuedRequestControllerMessenger(messenger),
+        messenger: buildQueuedRequestControllerMessenger(),
       };
       const controller = new QueuedRequestController(options);
+      // Trigger first request
+      const firstRequest = controller.enqueueRequest(
+        { ...buildRequest(), origin: 'https://exampleorigin1.metamask.io' },
```

Comment:
Nit: Since we have a factory function can we modify it to take in the same arguments that a request does and do this spreading for us?

Reply:
We could; I opted for this to keep the function simpler, in particular to avoid having to update the type signature so many times (it was written a few refactors ago)